### PR TITLE
Implement placement-in protocol for `Vec`

### PR DESCRIPTION
**Pull request by apasel422**
Open questions:
- Is `BackPlace` the right name for the struct?
- Should `&mut Vec` implement `Placer`? Or should it provide a `back_place` method [like `LinkedList` does](https://github.com/rust-lang/rust/blob/b53764c73bd722ea22142bace6249d5950066253/src/libcollections/linked_list.rs#L711) instead?

CC #30172, @rust-lang/libs, @pnkfelix

### CODE REVIEW

File: src/libcollections/vec.rs [hunk 1]
```diff
@@ -699,16 +699,7 @@ impl<T> Vec<T> {
...
-        // or if the length increment would overflow for zero-sized types.
-        if self.len == self.buf.cap() {
-            self.buf.double();
-        }
-        unsafe {
-            let end = self.as_mut_ptr().offset(self.len as isize);
-            ptr::write(end, value);
-            self.len += 1;
-        }
+        self <- value;
```
Comment (petrochenkov) [inline src/libcollections/vec.rs]: I'd avoid using placement in stable `push` without doing proper performance measurements.

Comment (nagisa) [inline src/libcollections/vec.rs]: I had some benchmarks implemented in https://github.com/nagisa/rust/commit/7b1b36da999139bbd2cff4b957f2042239ff6247 but they’re not enough at all to properly test all the cases (e.g. large-sized `T`).

**Comment by alexcrichton** [comment]
Nice! I'd be curious to discuss our conventions a little more here as well. It seems like we may want to both implement the placer on the types directly as well as have methods, but then we've got naming issues etc, etc. Fun bikesheds!

**Comment by aturon** [comment]
Yes, I think we should discuss the overall conventions here before going too much farther down that road -- let's be sure to cover it in the next libs meeting if we don't reach a resolution on this thread.

**Comment by bitshifter** [comment]
I'm very much a fly on the wall with this change. I was wondering if there's a precedent for an operator pushing into a container elsewhere? It looks a bit strange to me.

**Comment by petrochenkov** [comment]
>I think we should discuss the overall conventions here before going too much farther down that road 

The collection library conventions were specifically audited before 1.0. How on earth `Vec` ended up with `last()` as a way to obtain the last element and `VecDeque`/`LinkedList` with `back()`?
(Anyway, I still suggest the naming scheme mentioned in https://github.com/rust-lang/rust/pull/31696#issuecomment-184661573)

**Comment by petrochenkov** [comment]
Also, the PR name is a bit misleading, `v.place(i) <- elem` - the most interesting (implementation- and performance-testing-wise) part of placement protocol for `Vec` - is not implemented.

**Comment by nagisa** [comment]
@aturon when discussing the feature please keep in mind the following point:

This is unstable and implementation as it is currently is mostly done for everybody to try the feature and see what works and doesn’t. Notably the “Should &mut Vec implement Placer? Or should it provide a back_place method like LinkedList does instead?” should eventually answer itself through use – if people find `&mut Vec <- something` not pleasant to use, we would change to `.back_place` method before stabilisation.

> I was wondering if there's a precedent for an operator pushing into a container elsewhere? It looks a bit strange to me.

This is not an operator for pushing into containers; This is an operator for *directly* placing things into certain places that are not necessarily on the stack. Feature-wise the closest equivalent to this is C++’s emplacement, which, one could argue, hasn’t much special syntax.

**Comment by bitshifter** [comment]
> This is not an operator for pushing into containers; This is an operator for directly placing things into certain places that are not necessarily on the stack. Feature-wise the closest equivalent to this is C++’s emplacement, which, one could argue, hasn’t much special syntax.

From the example for this PR:

```
let mut vec = vec![1, 2];
&mut vec <- 3;
&mut vec <- 4;
assert_eq!(&vec, &[1, 2, 3, 4]);
```

Which is effectively the same as C++11 std::vector emplace_back() - maybe "push" isn't the right description? Placement or not, adding a new element to the end of the vector which sounds like a push operation to me.

**Comment by bluss** [comment]
Is there any way we can avoid this becoming a de facto deprecation of the `Vec::push()` method? I mean, if placement in is always better, it's going to be the preferred style?

**Comment by rkruppe** [comment]
@bluss It's always better when it works, but there are some theoretical circumstances where it doesn't work, e.g.:

```
&mut vec <- vec[0].clone();
```

More generally placement won't work if creating the value to be placed requires borrowing something also borrowed by the place. You can evaluate the value beforehand, of course, but this defeats the purpose of the placement syntax.

However, I'm not sure whether this will be common enough to warrant keeping `push` around. The `let value = vec[0].clone(); &mut vec <- value;` dance *shouldn't* have worse performance than `push`.

**Comment by petrochenkov** [comment]
When emplacement roll-back is expensive and happens often, `push` is also preferable:
```
vec.place(middle) <- if flip_coin() == HEAD { 1 } else { return };
```
Before flipping the coin vector shifts half of its elements to the right to prepare the scratch space for the new element in the position `i`. On `return` vector has to recover its initial state and shift these elements back.

**Comment by alexcrichton** [comment]
Thanks for the PR @apasel422! The libs team discussed this during triage yesterday, and the conclusion was that we probably want to hold off with more in-place implementations if we can. We're slightly uncomfortable that `LinkedList` already implements this, and we'd like to hold off on more if possible.

Specifically, especially as the discussion here indicates, we likely need to lay out some conventions for naming, usage, what implements `Placer`, etc. T. This should also manifest itself in the form of an RFC (to foster discussion at large), so in light of that I'm going to close this for now.

@aturon has showed interest in helping shepherd a conventions RFC in this regard, as well. @apasel422 would you be interested in writing such an RFC?

**Comment by apasel422** [comment]
@alexcrichton An RFC is totally reasonable for this (I'll start thinking about writing one in the next couple of days), but I am curious as to why #31696 was accepted.

**Comment by alexcrichton** [comment]
Yeah we concluded it was probably premature for us to accept https://github.com/rust-lang/rust/pull/31696 unfortunately :(

Thanks though!

**Comment by petrochenkov** [comment]
This is a very strange decision, emplacement needs implementation experience and performance measurements first of all, naming scheme is not important at the moment.

**Comment by aturon** [comment]
@petrochenkov @nagisa You both make reasonable points about wanting to experiment quickly and not being too cautious here. But, on the other hand, we'd prefer not to rapidly get into a situation where we have suboptimal de facto conventions or inconsistency across `std` -- which is what happened a lot prior to a dedicated API conventions process prior to 1.0.

(Personally, I think the names that [landed for `LinkedList`](https://github.com/rust-lang/rust/pull/31696) should be improved, and I'd rather that not accidentally become a difficult to change convention just because we didn't want to take the time to think about it.)

This doesn't need to be heavyweight -- the libs team would just like to see a brief writeup of some basic naming/impl scheme. We can move quickly on it.

**Comment by alexcrichton** [comment]
@apasel422 the libs team discussed this recently and our conclusion was that we have some concrete ideas of how to revive this movement of implementations of these traits on collections. Would you still be interested in tackling this?

**Comment by alexcrichton** [comment]
Ah and to be more concrete (and to write down the notes), our conclusion was:

* Collections like `Vec` which have a clear "default place" can implement the protocol directly, e.g. `vec <- item`.
* Otherwise standardizing around the convention `collection.place_location() <- item` seems reasonable. For example `vec.place_front() <- item`

**Comment by apasel422** [comment]
@alexcrichton Sure. I'll investigate resurrecting this later today.

**Comment by alexcrichton** [comment]
Awesome! We're also thinking that we could turn this into a tracking issue with all collections listed so we could keep up with the progress. Maybe you'd like to take a look first though and see where this would all need to be implemented?

**Comment by apasel422** [comment]
@alexcrichton I don't see how we can get syntax like

```rust
let mut vec = vec![1, 2, 3];
vec <- 4;
```

instead of

```rust
let mut vec = vec![1, 2, 3];
&mut vec <- 4;
```

as `Placer::make_place` takes self by value, and the placement-in syntax doesn't autoref.

**Comment by alexcrichton** [comment]
@apasel422 hm we were definitely not aware of that! We can perhaps push forward with `place_*` methods and handle that issue later, however.

**Comment by nagisa** [comment]
@apasel422 we could make auto-ref work by temoprarily changing desugaring from `Placer::make_place(var)` to `var.make_place()` and implemeting perma-unstable `impl Vec { fn make_place(&mut var) { Placer::make_place(var) }`.

Either way we have https://github.com/rust-lang/rfcs/issues/1286.

**Pull request closed without merging**